### ISSUE TITLE
A.CallTo is not side-effect free - can interfere with the usual default behaviour of a faked method

### ISSUE THREAD

**Issue by TimLovellSmith** [FakeItEasy/FakeItEasy#556]
Probably this topic has come up before but...

We're using FakeItEasy 2.0.0-beta009 and I am noticing a horrible pattern in our test code.
Unlike methods which return void that just silently finish by default, methods returning Task will return null by default and blow up if awaited.
Would there be a downside of having Fakes return a completed Task by default if the method return type is `Task` (and not `Task<T>`)?
If there is too much downside to changing `Fake<T>` what about introducing a new class 'Fake2<T>' like Fake but that which would have that behavior by default?

```C#
var createDeploymentCall = A.CallTo(() =>
    worker.CreateCacheDeployment(null, null, null, null, null, null, null, A<CancellationToken>._))
    .WithAnyArguments();
createDeploymentCall.Returns(Task.Delay(0));
var deleteStageSlotCall = A.CallTo( () => azureTasks.DeleteStageSlotAsync(null, null, A<CancellationToken>._)).WithAnyArguments();
deleteStageSlotCall.Returns(Task.Delay(0));
var swapDeploymentCall = A.CallTo(() => azureTasks.SwapDeploymentAsync(null, null, null, A<CancellationToken>._)).WithAnyArguments();
swapDeploymentCall.Returns(Task.Delay(0));
```

**Comment by adamralph** [comment]
Hi @TimLovellSmith 

Fake methods have returned a completed task by default since 0.13 - see https://github.com/FakeItEasy/FakeItEasy/issues/129

Can you show an example which is not displaying this behaviour?

**Comment by TimLovellSmith** [comment]
Oh. :D
I just discovered the same by writing a unit test. I have some code to clean up!

**Comment by blairconrad** [comment]
Ah. I thought the confusion might've been between `Task` and `Task<T>` - @TimLovellSmith's original comment referenced both but without backticks, so they both looked like "Task".
Glad it's been resolved.

**Comment by blairconrad** [comment]
While I'm fixing things (no change is too small!) corrected https://github.com/FakeItEasy/FakeItEasy/issues/556#issuecomment-149961907 to indicate the "completed Task" behaviour was added in 1.13, not 0.13 (which I think wasn't a thing).

**Comment by adamralph** [comment]
Oops, thanks :wink:

**Comment by TimLovellSmith** [comment]
So... actually reopening as this needs deeper digging. Fake *does* sometimes return a completed Task by default... and it also sometimes does *not*.

Here is a case where it seems to work:

        [Test]
        public void Should_return_completed_task_for_task_method()
        {
            // Arrange
            var fake = A.Fake<IFoo>();
            // Act
            Task t = fake.BarAsync();
            // Assert
            Assert.NotNull(t);
            t.Wait();
        }

And here is a case where it does not:



        [Test]
        public void Should_return_completed_task_for_task_method()
        {
            // Arrange
            var fake = A.Fake<IFoo>();

            // Act
            var theCall = A.CallTo(() => fake.BarAsync()).WithAnyArguments();
            Task t = fake.BarAsync();

            // Assert
            Assert.NotNull(t);
            t.Wait();
            theCall.MustHaveHappened();
        }

As you can see, the act of 'choosing' a call to work with will break its default return value behavior.

**Comment by adamralph** [comment]
```C#
var theCall = A.CallTo(() => fake.BarAsync()).WithAnyArguments();
```

is an incomplete statement and achieves nothing. It should be followed by 

```C#
theCall.Returns(...); // where ... is some task that you want to control
```

Incidentally I'm not sure why you are using this pattern of assigning the call to a variable and then adding a return value to it? Why not just

```C#
A.CallTo(() => fake.BarAsync()).WithAnyArguments().Returns(...);
```

**Comment by TimLovellSmith** [comment]
@adamralph It is so that we can write theCall.MustHaveHappened(); as an assertion at the end

**Comment by adamralph** [comment]
OK, fair enough. I almost never have the need to verify a call so I tend to forget about that.

But still, in the example you've given, you are not setting up a return value so the reference to the call is way too early, i.e. it should happen only in the assertion.

```C#
// Assert
Assert.NotNull(t);
t.Wait();
A.CallTo(() => fake.BarAsync()).WithAnyArguments().MustHaveHappened();
```

**Comment by TimLovellSmith** [comment]
Great that there's a workaround.
But should doing A.CallTo() without any return value configuration actually mess with the return value behavior? It gets kind of inconvenient if you have to move the code around when you switch it from having a return value to not having one but still having an assertion.

**Comment by blairconrad** [comment]
I see both points. The intent of FakeItEasy's API, as I understand it, is to provide a nice, fluent interface.
Assigning the result of `A.CallTo` circumvents that.
On the other hand, I've done it a number of times when I've had to configure a method and also wanted to verify a call. It's quite convenient.
On the gripping hand, I've never done it when I haven't been both configuring and verifying a method call. Never occurred to me, and I probably wouldn't, as I prefer to read the fluent sentence and would rather not "waste" a variable.
… I'm not sure what comes after "gripping hand". Anyhow, just because _I_ wouldn't do it doesn't mean it shouldn't be done. I can understand why @TimLovellSmith might prefer a consistent way of dealing with configuring and/or verifying calls.

And the current behaviour violates the Principle of Least Astonishment. I think it's preferable for `A.CallTo(()=>…)` to have no side effect.

I'm inclined to rename this issue to describe the generic problem, that invoking `A.CallTo` can interfere with the usual default behaviour of a faked method, argue for a few days over whether fixing it is an enhancement or a bug fix, and add it to the backlog.

**Comment by blairconrad** [comment]
If we do attempt alter the behaviour of a naked `A.CallTo`, we might consider #534 as well, which highlights that if the result of `A.CallTo` is saved away and configuration specified upon it more than once, the behaviours don't stack up, but rather override.
The issues may be addressable independently (assuming we want to address either or both), but good to keep them both in mind.

**Comment by adamralph** [comment]
And the current behaviour violates the Principle of Least Astonishment. I think it's preferable for A.CallTo(()=>…) to have no side effect.

:+1: 

I'm inclined to rename this issue to describe the generic problem (that invoking A.CallTo can interfere with the usual default behaviour of a faked method—it's not really a Task-related problem), argue for a few days over whether fixing it is an enhancement or a bug fix, and add it to the backlog.

:+1: 

If we can solve #534 at the same time then great, but not necessary.

**Comment by blairconrad** [comment]
> If we can solve #534 at the same time then great, but not necessary.

Sure, but I think these are separable.
I mean they _could_ be fixed at the same time, but I bet the changes would be different enough that we'd have separate commits that could as happily live in separate PRs. But I'm just guessing.

**Comment by blairconrad** [comment]
While we're at it, we should take a gander at `NextCall.To`.

**Comment by thomaslevesque** [comment]
I'm resurrecting this issue, because I think it's important. I started making some changes, which are quite simple, but I have a few failing tests that I don't really understand yet.

**Comment by thomaslevesque** [comment]
I marked it as breaking, because it will change the current behavior when calling and not using `A.CallTo`. Probably not a very big breaking change, since the current behavior doesn't really make sense anyway.

**Comment by adamralph** [comment]
If it's breaking, then it should be marked `on-hold`, unless we want to take the plunge and shoot for 3.0.

**Comment by adamralph** [comment]
This issue also never made it in to `ready` so I guess we don't really have consensus on what should happen here.

**Comment by blairconrad** [comment]
I wouldn't call this breaking. Calling `A.CallTo` (and friends) with no further configuration is an abuse (or at least misuse) of the API.
I know, every bug is someone's favourite feature, but I can't consider that `A.CallTo` to changing a return value to something we would not otherwise return one of them.

**Comment by blairconrad** [comment]
As for what should happen, I think we all agree at least up to the point that `A.CallTo` should be free of side effects. But you're right in that there's still some room for interpretation.

I was just about to comment on @thomaslevesque's implementation. I think it does a fine job of removing the side effect.
When this issue first came up, I made a comment about considering #534 at the same time. In my mind (and I shouldn't've kept it there), I'd envisaged a solution where:

```c#
var aCallToBar = A.CallTo(() => foo.Bar());
aCallToBar.Returns(1);
aCallToBar.Throws<Exception>().Once();
```

Would cause calls to `Bar` to throw once and then return `1` thereafter, just as if we'd inlined `aCallToBar`.

(I think the implementation would be similar to @thomaslevesque's in #840, except that we'd introduce an intermediate object that the existing `RuleBuilder` would create from itself and attach to the manager when needed. The new object would support all the remaining actions that the original `RuleBuilder` would, but would not reattach itself to the manager. In some ways, this is not dissimilar to portions of @thomaslevesque's `Then` implementation, with its call specification-cloning.)

In this way, we remove the side effect as well as the restriction on not using `aCallTo` two specify (for example) `Returns` twice. Of course, we may soon have a much better way to specify changing behaviour in #804.

Anyhow, I hadn't written anything earlier about my idea because I wasn't sure that it would work (I _think_ so), and hadn't had a chance to spike it.

Of course, merging the final version of #840 doesn't preclude such an approach later, if we think the behaviour I describe is desirable.

**Comment by thomaslevesque** [comment]
> If it's breaking, then it should be marked on-hold, unless we want to take the plunge and shoot for 3.0.

Ah, yes. I added the label. Does it make sense to have both `on-hold` and `in-progress`? I guess the `on-hold` just means that the PR won't be merged immediately, but doesn't prevent working on it?

> This issue also never made it into ready so I guess we don't really have consensus on what should happen here.

True, the label is missing. But I was under the impression that all participants agreed on the fact that `A.CallTo` should be side-effect free.

> I wouldn't call this breaking. Calling A.CallTo (and friends) with no further configuration is an abuse (or at least misuse) of the API.
> I know, every bug is someone's favourite feature, but I can't consider that A.CallTo to changing a return value to something we would not otherwise return one of them.

Well, it would break badly written tests that work only because of the current behavior, so technically, I guess it's breaking. But at least it's a breaking change that probably no one would argue against :wink:

**Comment by adamralph** [comment]
The state transition labels (`ready` and `in-progress`) only have meaning in a collaborative sense. (Note that they represent state transition, specifically on the context of https://waffle.io/FakeItEasy/FakeItEasy so it doesn't make sense to have both labels.) Individually we can be working on whatever we want. :smile:

We don't have anything in https://github.com/FakeItEasy/FakeItEasy/wiki/Issue-labels to describe exactly what they mean, but my understanding is that `ready` means we have consensus on what needs to happen when the issue moves into `in-progress` in order to close it.

> Does it make sense to have both `on-hold` and `in-progress`?

I'm not sure that it does. For me `on-hold` means we have a reason not to do this work now. As I said above, if you want to start a specualtive WIP PR as an individual, no one is going to stop you, but collaboratively we are holding off from working on this for some reason.

**Comment by adamralph** [comment]
In the case of this issue, I think we should move the issue back to the backlog (by removing the `in-progress` label) and mark it `ready` when we're sure we have consensus on how/if the bug should be fixed.

**Comment by thomaslevesque** [comment]
> In the case of this issue, I think we should move it back to the backlog (by removing the in-progress label) and mark it ready when we're sure we have consensus on how/if the bug should be fixed.

OK, I removed the label.

What do you think is missing to mark it `ready`?

Goal: the `A.CallTo` methods should be made pure (i.e. no side effects, and the output depends only on the inputs). They should have no effect on the fake's behavior until we actually specify a behavior.

A way to achieve this (which I started to implement) is to:
- not add the rule to the fake manager while no behavior has been specified
- add the rule to the fake manager as soon as a behavior is specified.

**Comment by blairconrad** [comment]
> What do you think is missing to mark it `ready`

To decide on a general path for the implementation, probably.
As far as the change in behaviour, I'm more than content with the path you're taking.
I say carry on with your implementation.
Of course, if the consensus is "breaking", we'd hold off for a major release, no?

I did make [my comment](https://github.com/FakeItEasy/FakeItEasy/issues/556#issuecomment-242966624) about an alternative that I'd imagined, just as something to consider.

If we think that my option is a bad idea and we prefer the current behaviour of throwing exceptions when e.g. `Returns` is called twice, then I'll drop it.
If we think it's complementary and can be addressed later, let's spin off another issue that can be addressed in the same or a later PR.

Suits?

**Comment by thomaslevesque** [comment]
> I'd envisaged a solution where:
> 
>     var aCallToBar = A.CallTo(() => foo.Bar());
>     aCallToBar.Returns(1);
>     aCallToBar.Throws<Exception>().Once();
> 
> Would cause calls to `Bar` to throw once and then return 1 thereafter, just as if we'd inlined `aCallToBar`.

I like it, I think. My only worry is that we first prevented doing this by throwing an exception, and now we re-enable it with different semantics. People upgrading from 1.x to whatever version includes this change will get a different behavior if they already had code like this.

**Comment by blairconrad** [comment]
> My only worry is that we first prevented doing this by throwing an exception, and now we re-enable it with different semantics. People upgrading from 1.x to whatever version includes this change will get a different behavior if they already had code like this.

It is a concern. Of course the major version bump does give us that license.

Also, I was thinking about this while on the exercycle and maybe it's not worth doing. The implementation you're working on is undoubtedly simpler, and once we get your vastly superior `.Then` API in, the I think my example would becoem

```c#
var aCallToBar = A.CallTo(() => foo.Bar());
aCallToBar.Throws<Exception>().Once()
          .Then.Returns(1);

…

aCallToBar.MustHaveHappened();
```

(since the only reason to save the call specification in a variable would be to assert it)

And that's just so much better than what I'd had!

**Comment by adamralph** [comment]
I understand the motivation to make `A.CallTo()` pure and I think it's the right thing to do.

What is the breaking change that is causing concern?

**Comment by thomaslevesque** [comment]
> What is the breaking change that is causing concern?

Consider a test with this code:

    A.CallTo(() => foo.Bar()).Returns(42);
    A.CallTo(() => foo.Bar());

Currently, `foo.Bar()` will return 0. If we fix this issue, it will return 42. This is certainly more correct, but it will still break some badly written tests.

**Comment by adamralph** [comment]
Personally I'd call that a non breaking change. The current behaviour makes
no sense at all and I would consider it incorrect and a bug. Those "badly
written" tests which are relying on this behavior are bad enough that they
should be changed immediately and I think it's good that an update to a new
patch version (or minor) will inform the authors of that fact.

As we've discussed before, *any* change is potentially breaking so it's a
subjective call, but I vote to call this non-breaking.

**Comment by thomaslevesque** [comment]
> Those "badly written" tests which are relying on this behavior are bad enough that they should be changed immediately and I think it's good that an update to a new patch version (or minor) will inform the authors of that fact.

I like this approach :smile:

OK, I'm convinced. Let's remove the breaking tag.

**Comment by TimLovellSmith** [comment]
@adamralph @thomaslevesque I hear you seem to be arguing about breaking or nonbreaking. FWIW I think it *is* breaking, just by a common sense test: if people upgrade to your new version of the library and a lot of tests break, that should be a breaking change.

A breaking change should clearly go in v3. Are you also wondering when to start v3?

Are there other bugs worth fixing in v2? If so, I don't see any reason not to postpone this A.CallTo change a little longer - it may be a 'buggy' behavior in terms of my expectations, but once you understand it and write your tests accordingly it doesn't block you from getting any work done (which I think is something that your developer audience cares about a lot).

[PS Apologies for the verbosity.]

**Comment by adamralph** [comment]
Since we're agreed on the approach, and that it's non-breaking, and @thomaslevesque has started work, I've moved back into `in-progress` and removed `on-hold`.

**Comment by adamralph** [comment]
@TimLovellSmith it's not quite as simple as that. You can argue that [every public change is a breaking change](https://blogs.msdn.microsoft.com/ericlippert/2012/01/09/every-public-change-is-a-breaking-change/). This means "breaking" is down to interpretation, project policy, and sometimes case by case opinion (as in this case).

**Comment by blairconrad** [comment]
@TimLovellSmith, it's just as @adamralph says. And one more qualification 

> if people upgrade to your new version of the library and a lot of tests break, that should be a breaking change.

I expect very few tests are taking advantage of the current (incorrect) behaviour. I know you've identified these in your own codebase, but it seems very weird, and I'd definitely steer people away from using the pattern as you've done.

If you're concerned about your existing tests, now's a good time to pick up the [FakeItEasy Analyzer](https://www.nuget.org/packages/FakeItEasy.Analyzer). It'll tell you today where you might have a problem with this pattern in the future!

**Comment by blairconrad** [comment]
Sorry, @TimLovellSmith. You'd also asked

> Are there other bugs worth fixing in v2?

[There are. And enhancements to be made too.](https://github.com/FakeItEasy/FakeItEasy/issues?utf8=%E2%9C%93&q=is%3Aopen%20no%3Aassignee%20label%3Aready%20-label%3Aon-hold)Take your pick!

**Comment by TimLovellSmith** [comment]
@adamralph Lippert's blog is an interesting proof that every change [to public surface area] can be a breaking change - in the sense of proving some trivia that is at least interesting to e.g. mathematicians, compiler writers, and perhaps library maintainers. But he concludes with:

> [it is] such an obviously contrived and unlikely scenario that no “real world” developers are likely to run into it. When we are evaluating the impact of potential breaking changes on our customers, we now explicitly discount this flavour of breaking change as so unlikely as to be unimportant

It is partly admitted in this thread that we could be discussing the opposite to Eric's example of a *hypothetically breaking* change (OK, thank you mathematicians, *provably breaking*), that doesn't actually break anybody in practice. I.e. this could actually break someone's test in practice.

While I originally raised the bug because I didn't like the way things work, I feel a guilty (for having raised the issue?) need to defend anybody who relies on this existing behavior. Why should my need for least surprise trump their need for tests to keep working, especially when they expected their tests to work even while doing a 'minor' package upgrade to get some other bugfix from that wonderful list?

**Comment by blairconrad** [comment]
This issue has been released in [FakeItEasy 2.3.0](https://github.com/FakeItEasy/FakeItEasy/releases/tag/2.3.0).

@thomaslevesque, thanks again, and look for your name in the release notes! :trophy:
